2주차 - Spring Web을 이용한 To Do REST API 만들기

steve7867 opened the pull request:
이번 주도 잘 부탁드립니다~~

steve7867 committed:
Spring dev tools 의존성 추가, Task, TaskIdGenerator, TaskRepository 생성 및 구현

steve7867 committed:
TaskDto, TaskService,TaskServiceImpl 추가

steve7867 committed:
Task 생성자 추가

steve7867 committed:
TaskController 추가

steve7867 committed:
예외 처리 추가

steve7867 committed:
TaskIdGenerator, TaskRepository 인터페이스 추가

TaskIdGenerator, TaskRepository 인터페이스를 추가하고 기존의 TaskIdGenerator, TaskRepository는 각각 TaskIdGeneratorImpl, TaskIdGeneratorImpl로 변경하고 두 인터페이스를 구현함

steve7867 committed:
의존성 주입 방식 변경

필드 주입에서 생성자 주입 방식으로 변경

steve7867 committed:
@CrossOrigin 추가

steve7867 committed:
TaskController postTask() 수정, TaskRepositoryImpl findAllTasks() 수정

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -0,0 +1,51 @@
+        return task;
+    }
+
+    public Task changeTitle(Long id, String newTitle) {
+        final Task originalTask = taskMap.get(id);
+        if (originalTask == null) {
+            return null;
+        }
+
+        originalTask.setTitle(newTitle);
+        return originalTask;
+    }
```
`null`을 "값 없음"의 의미로 사용하고 있습니다. 이번에는 `Optional`을 여기에 적용해 보면 어떨까요?

steve7867 replied:
`Optional`을 적용했습니다. `Optional`을 사용하니까 어떤 메소드의 리턴값이 `null`일 수도 있는지 파악하기가 용이하고 코드도 짧아지네요.

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -0,0 +1,51 @@
+        return originalTask;
+    }
+
+    public Task deleteById(Long id) {
+        return taskMap.remove(id);
+    }
```
추가 과제로 여러 id 를 받아서 한꺼번에 삭제하는 메소드도 작성해 봅시다.

steve7867 replied:
요구 사항을 더 구체적으로 말씀해주시면 안 될까요?😭

johngrib replied:
여러 아이디를 받아서 삭제하면 됩니다. 메소드 시그니처 후보로 이런 것들을 떠올릴 수 있겠죠.

- `public void deleteTasks(List<Long> taskIdList)`
- `public int deleteTasks(Collection<Long> taskIds)`
- `public DeleteReport deleteTasks(Set<Task> tasks>`
- ...

가능한 경우는 엄청나게 많습니다.

steve7867 replied:
```java
public Map<Long, Task> deleteTasks(List<Long> idList) {
    final Map<Long, Task> removedTasks = new ConcurrentHashMap<>();

    for (Long id : idList) {
        Task removedTask = taskMap.remove(id);
        removedTasks.put(id, removedTask);
    }

    return removedTasks;
}
```

이렇게 구현해봤습니다.
리턴 타입을 `Map<Long, Task>`로 한 이유는 요청된 id의 Task가 삭제되었는지 확인할 수 있게 하기 위함입니다.
리턴된 map은 요청된 id가 삭제되었다면 삭제된 Task를 반환하고 요청된 id의 Task가 Repository에 없었다면 null을 반환합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/TaskDto.java:
```diff
@@ -0,0 +1,31 @@
+package com.codesoom.assignment.models;
+
+public class TaskDto {
+
+    private Long id;
+    private String title;
+
+    public TaskDto() {
+    }
+
+    public TaskDto(Task task) {
+        this.id = task.getId();
+        this.title = task.getTitle();
+    }
+
```
TaskDto has a Task 의 관계가 되도록 이 클래스를 수정해보는 연습을 해봅시다.
(Decorator 패턴을 조사해보시고, TaskDto가 Task의 decorator가 되도록 하면 됩니다.)

steve7867 replied:
![UML](https://user-images.githubusercontent.com/115217247/196419964-33455a8f-8afe-4f14-856c-55fc0fcc4214.png)

이렇게 구현했습니다.
Decorator 패턴을 무리하게 적용한 것 같습니다.
Decorator class를 제거하고, TaskDto가 Task를 직접 필드로 가지도록 만드는 게 더 나을 것 같네요.
궁금한 게 있는데 실무에서 Dto와 Entity를 Decorator 패턴으로 사용하는 경우가 많나요?
`Decorator 패턴은 기능을 동적으로 추가하는 게 목적` 인 디자인 패턴인데 Dto와 Entity에 Decorator 패턴을 적용하는 경우가 어떤 경우인지 궁금합니다.

johngrib replied:
이렇게 강력한 실행력으로 바로 작업해버리실 줄은 몰랐어요... 말씀하신대로 `Decorator` 클래스를 제거하는 것이 나아 보입니다.
실무에서는 이렇다 저렇다 하기가 어려운게 사람마다 다 다르게 합니다. 물론 대부분의 경우는 DTO와 Entity를 철저하게 격리하죠. 그런데 DTO가 Entity가 지시하는 정보의 출력 포맷을 담당한다는 걸 생각해보면 DTO가 Entity를 데이터 소스로 몰래 갖고있게 하고 적절한 getter를 제공하는 방법이 유용할 때가 드물지 않게 있습니다. 흔히 쓰지 않을 뿐이죠. 이걸 기억해두시면 필요한 순간에 사용하실 수 있을 거에요.

steve7867 replied:
Decorator 패턴이 신기해서 써보고 싶어서 바로 적용해 봤습니다 ^^
말씀하신 대로 DTO has a Entity관계가 어떤 경우에 유용할 수 있는지 기억해두겠습니다.

johngrib replied:
좋은 공부가 되었길 바랍니다. 한편 최근에 다른 분의 PR에 남긴 DTO에 대한 코멘트도 함께 읽어보시면 도움이 될 수 있을 것 같아요.

- https://github.com/CodeSoom/spring-week5-assignment-1/pull/79#discussion_r995706395
- https://github.com/CodeSoom/spring-week5-assignment-1/pull/79#discussion_r994584084

보통 나중 주차 코멘트는 잘 소개해드리지 않는데, @steve7867 님은 열심히 고민하며 잘 진행해나가고 계셔서 소개해 드리면 시너지가 발생할 것 같다는 생각이 들어서 소개해드립니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,28 @@
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
```
추가 과제: Task 클래스가 setter를 제공하지 않아야 한다는 제약이 주어졌습니다. setter를 삭제하고 대안이 될 수 있는 방법을 생각해서 적용해 보세요.

steve7867 replied:
```java
public BaseTask addTask(TaskDto dto) {
    dto.setId(idGenerator.allocateId());
    final BaseTask task = dto.toTask();
    taskMap.put(task.getId(), task);
    return task;
}
```
새로운 Task를 생성하는 작업은 위와 같이 구현했습니다.
Task와 TaskDto에 Decorator 패턴을 적용하다보니 dto가 Repository 영역까지 침투했네요.
task의 id를 할당하는 책임을 Service 영역에 위임해도 괜찮을까요?

```java
public Optional<BaseTask> changeTitle(BaseTask task) {
    if (!taskMap.containsKey(task.getId())) {
        return Optional.empty();
    }

    taskMap.put(task.getId(), task);
    return Optional.of(task);
}
```
기존의 Task의 title을 변경하는 작업은 위와 같이 구현했습니다.

johngrib replied:
음 이렇게 하기보다는 Task의 생성을 책임지는 Task 생성기를 만들어보는 것도 고려해 보세요. id 생성의 책임도 Task 생성기가 떠맡으면 적절할 수 있겠죠.

steve7867 replied:
`TaskGenerator` 클래스를 따로 빼서 Task 생성 책임을 위임했습니다.

johngrib reviewed (commented):
이번주도 화이팅!

steve7867 committed:
Optional 적용 및 관련 코드 리팩토링

steve7867 committed:
Task class setId(), setTitle() 제거 및 관련 코드 리팩토링

setter를 제거함으로써 Task를 불변 객체화

steve7867 committed:
Task, TaskDto에 Decorator 패턴 적용

steve7867 committed:
Decorator 적용 관련 리팩토링

steve7867 committed:
or() 메소드를 이용한 코드 간소화

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Decorator.java:
```diff
@@ -0,0 +1,28 @@
+package com.codesoom.assignment.models;
+
+public abstract class Decorator extends BaseTask{
```
`Decorator` 클래스를 정의하는 것은 과도한 작업 같습니다.

steve7867 replied:
제가 생각해도 너무 억지로 Decorator 클래스를 만들었네요.
단순한 TaskDto has a Task 관계로 수정했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/TaskDto.java:
```diff
@@ -1,31 +1,38 @@
 package com.codesoom.assignment.models;
 
-public class TaskDto {
+public class TaskDto extends Decorator {
```
상속을 사용하지 않아도 괜찮습니다. 이 작업의 목표는 has a 관계를 구현하는 것을 연습하는 것이기 때문입니다.

steve7867 replied:
수정했습니다

johngrib commented on app/build.gradle:
```diff
@@ -35,6 +35,9 @@ dependencies {
     implementation 'com.fasterxml.jackson.core:jackson-databind:2.11.3'
 
     // Spring Web
+    developmentOnly 'org.springframework.boot:spring-boot-devtools'
```
이렇게 build.gradle 파일에 새로운 dependency를 추가하는 경우에는 코드 리뷰의 형태로 설명을 달아보세요. 코드리뷰하는 동료들이 자료를 조사하는 수고를 덜어주게 될 겁니다.

steve7867 replied:
혹시 예시를 들어주실 수 있으실까요?

그런데 제가 주석을 엉뚱하게 달아 놓았네요 ㅎㅎ
Spring devtools에 'Spring Web'을 적어 놨네요.
수정하겠습니다~

johngrib replied:
지금 우리가 PR에 코멘트를 달면서 이야기하는 것처럼 하는 거죠! 즉 코드 리뷰어보다 먼저 자기 자신이 이 라인에 코멘트를 달아서 "이건 이러이러한 목적으로 추가한 디펜던시인데, 이러저러한 기능이 있고 프로젝트에 도입하면 이런 효과가 있어서 추가했습니다" 처럼 하는 거죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,76 @@
+
+    private final TaskService service;
+
+    @Autowired
```
이건 생략할 수 있습니다.

steve7867 replied:
> If a class only declares a single constructor to begin with, it will always be used, even if not annotated. An annotated constructor does not have to be public.

공식 문서에도 생략 가능하다고 나오네요!
꿀팁 감사합니다~

johngrib replied:
네 이 방식이 스프링을 만든 사람들이 권장하는 방법이기도 합니다.

steve7867 committed:
build.gradle 주석 수정, @Autowired 제거

steve7867 committed:
TaskDto has a Task 관계로 변환

steve7867 committed:
TaskGenerator, TaskRepositoryImpl 추가

steve7867 committed:
관련 코드 리팩토링

steve7867 committed:
Task 클래스 regDate, modDate 필드 추가 및 TaskDto setter, getter 수정

steve7867 committed:
TaskGenerator 메소드 수정 및 관련 코드 리팩토링

steve7867 committed:
deleteTasks() 추가

steve7867 committed:
TaskGeneratorImpl 각 메소드에 @Override 추가, TaskRepository 명세 변경, TaskGeneratorImpl 버그 수정

steve7867 committed:
Task 객체 생성 방식과 title 변경 방식 수정 및 관련 코드 리팩토링

steve7867 committed:
TaskDto setter를 Chaining 방식으로 변경

steve7867 committed:
Task getter 접근 제어자 변경

default -> public

johngrib reviewed (commented):
github을 사용할 때에는 `@` 를 주의하셔야 합니다. 이렇게 다른 github 사용자를 링크하게 되거든요. 이러면 그 분에게 알림이 가게 됩니다. 여기까지 보러 오시게 될 수도 있겠죠.

![image](https://user-images.githubusercontent.com/1855714/196688461-ad6db2b3-894d-4b29-8dd0-863b3ffe9a26.png)

그러니 이런 경우에는 ``` ` ``` 을 양쪽에 붙여 주시면 됩니다.
```
`@Autowired`
```

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    public Map<Long, Task> deleteTasks(List<Long> idList) {
```
이렇게 하는 것도 고려할 수 있겠습니다.
```suggestion
    public Set<Long, Task> deleteTasks(Set<Long> idList) {
```

steve7867 replied:
```java
public Set<Task> deleteTasks(Set<Long> idList) {
```
이런 방식을 말씀하시는 게 맞을까요?

johngrib replied:
네. 맞습니다. 리턴 타입은 제가 오타를 냈네요! 다만 고려할 점들이 몇 가지 더 있어요.

- `idList`: 이제 List 가 아니라 Set을 사용하게 되었습니다.
- 리턴 타입: Set 으로 충분하지 않을 수도 있습니다. 만약 Task가 아주 중요한 것이라면 (예를 들어 금융 기관에서의 데이터 마이그레이션 Task 일 수도 있습니다) 삭제 결과로 좀 더 복잡한 응답을 기대할지도 모릅니다. 각각의 Task가 삭제된 시간과 Task 백업 아이디 같은 것들이 포함된 복잡한 타입 같은 거 말이죠. 이런 경우엔 `DeleteReport` 같은 타입을 만들어 리턴하는 것도 고려할 수 있습니다. 다행히 지금 우리가 하고 있는 과제는 단순한 Todo 이고 이런 작업들이 불필요하므로 `Map`, `Set`또는 삭제된 Task의 카운트를 의미하는 int 정도로도 만족할 수 있습니다. 그러나 복잡하고 중요하고 위험한 정보를 다루게 된다면 이런 식으로도 작업할 수 있다는 것을 기억해두면 좋을 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -0,0 +1,51 @@
+    public Map<Long, Task> deleteTasks(List<Long> idList) {
+        final Map<Long, Task> removedTasks = new ConcurrentHashMap<>();
+
+        for (Long id : idList) {
+            Task removedTask = taskMap.remove(id);
+            removedTasks.put(id, removedTask);
+        }
```
removedTasks에 하나 하나 put하는 건 꽤 괜찮아보이지만 함정이 하나 있습니다. 만약 다음과 같이 아이디가 주어진다고 생각해 봅시다.


- 1 : 이미 과거에 삭제된 Task의 아이디
- 2 : 현재 repository에 저장되어 있는 Task의 아이디
- 3 : 아직 등록되지 않은 미래의 Task 아이디

이렇게 된다면 실제로 삭제되는 Task는 2번일 것입니다. 하지만 리턴되는 removedTasks 에는 1, 2, 3이 모두 들어있지 않을까요?

steve7867 replied:
맞습니다.
deleteTasks()의 파라미터를 Set<Long>타입으로 변경해주고, null 체크 로직을 추가해서 문제를 해결했습니다.
기능을 너무 허술하게 만들었다는 생각이 드네요.
항상 기능을 만들면 테스트 코드를 짜서 검증을 해야겠습니다.

johngrib replied:
훌륭한 결론입니다. 항상 테스트 코드를 작성해서 코드로 충분히 표현되지 않은 중요한 특징들을 명시해두면 좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+    private final Map<Long, Task> taskMap = new ConcurrentHashMap<>();
+
+    public Collection<Task> findAllTasks() {
+        return Collections.unmodifiableCollection(taskMap.values());
+    }
```
만약 가장 최근에 등록된 Task 100개를 최신순으로 정렬된 리스트를 리턴하는 메소드가 필요하다면 어떻게 할 수 있을까요?

steve7867 replied:
Brute-force 방식으로 일단 구현했습니다.
메소드를 호출할 때마다 최근에 등록된 task를 100개씩 찾아야 한다는 문제가 있습니다.
최적화 방법을 고민해보겠습니다.

johngrib replied:
좋습니다. 메소드를 하나 추가해보는 것을 목표로 구상해 보도록 합시다. 테스트코드를 먼저 만들어봐도 괜찮을 겁니다.

steve7867 replied:
만약 메소드를 호출하는 쪽에서 최근에 추가된 task를 항상 100개만 필요로 한다면 100개의 최근 추가된 task를 따로 저장해 두는 방법(캐시 기법)을 사용할 수 있을 것 같습니다.
(항상 100개만 요구한다고 가정하고 100개를 항상 캐시하는 방법으로 구현을 시도했습니다만 다른 메소드들까지 건드리게 되어 코드가 상당히 지저분해지고 가독성이 떨어지는 것을 느꼈습니다.)
그런데 100개, 500개, 1000개, 200개처럼 필요로 하는 개수가 들쭉날쭉하다면 캐시 기법을 사용하기 어려운 것 같습니다. 만약 메소드 사용에 관한 데이터가 쌓여서 평균적으로 몇 개의 task를 요구하는지를 알고 있다면 그것보다 조금 더 넉넉한 개수만큼 캐시를 해뒀다가 요청되는 개수에 맞춰서 추가 작업을 해준 뒤 호출부에 반환을 해줄 수 있을 것 같습니다.
메소드 사용에 관한 데이터가 없으므로 우선 Brute-force한 방식을 유지했습니다만 트래픽이 늘어난다면 당연히 성능 문제가 발생하지 않을까하는 생각이 듭니다.

johngrib replied:
이런 경우에는 별도의 서비스 클래스를 작성하는 것도 고려해볼만 합니다. 캐시를 사용하는 아이디어는 괜찮아 보입니다. 시간과 여유가 된다면 구현해 보는 것도 좋을 것 같네요.

steve7867 replied:
몇 개 정도를 캐시해두는 것이 좋을까요?

steve7867 replied:
일단 최근에 추가된 task들을 캐시하는 책임을 가진 RecentlyAddedTasksFinder를 추가했습니다.
현재는 TaskRepositoryImpl과 RecentlyAddedTasksFinder를 별도의 클래스로 구현을 했습니다만 RecentlyAddedTasksFinder가 TaskRepositoryImpl의 decorator가 되도록 구현하는 것이 더 괜찮을 것 같다는 생각이 듭니다.
왜냐하면 RecentlyAddedTasksFinder는 TaskRepositoryImpl의 기본 기능에 더해 최근에 추가된 task들을 찾아주는 추가 기능의 역할을 하기 때문입니다.

steve7867 replied:
![제목 없는 다이어그램 drawio](https://user-images.githubusercontent.com/115217247/197388851-06e0e102-1c8b-4bd7-ab80-eab6bf91c971.png)

Decorator 패턴 보다는 상속 관계를 가지는 게 `코드가 더 단순하고 가독성이 좋지 않을까`하는 생각이 들어 상속 관계를 가지는 형태로 구현했습니다.
캐시 사이즈는 임의로 300개로 했습니다(몇 개가 적당한지 판단하기가 어렵네요)
메소드 호출부에서 캐시 사이즈보다 많은 개수를 요구하면 이전에 추가된 task들을 더 추가하고, 더 적은 개수를 요구하는 경우에는 요구하는 개수에 부합하는 headMap을 찾는 방식으로 구현했습니다.
코드가 많이 지저분하네요.....

johngrib commented on app/src/main/java/com/codesoom/assignment/models/TaskDto.java:
```diff
@@ -0,0 +1,45 @@
+        if (task == null) {
+            return;
+        }
+        setId(task.getId()).setTitle(task.getTitle());
```
메소드 체인 기법을 잘 사용하셨습니다.

steve7867 committed:
deleteTasks() 버그 수정 및 테스트 코드 작성

steve7867 committed:
findRecentlyAddedTasks() 추가 및 테스트 코드 작성

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -3,22 +3,43 @@
+        final List<Task> recentlyAddedTaskList = new ArrayList<>();
+
+        for (Task task : taskMap.values()) {
+            if (recentlyAddedTaskList.size() >= 100) {
```
100 이 메소드 인자로 제공되면 더 좋을 겁니다. 지금은 100개만 필요하지만 사용자 디바이스나 출력 요구에 따라 다른 값들이 필요할 수도 있겠죠.

- 모바일 앱: 100개의 Task 조회
- 프린터기: A4 용지의 크기에 맞는 수가 프린터 드라이버에서 제공됨
- 웹 사이트(고객): 200개의 Task 조회
- 웹 사이트(관리자): 1000 개의 Task 조회

steve7867 replied:
기본적인 건데 제가 놓쳤네요

johngrib replied:
그럴 수도 있죠 ㅎㅎ

steve7867 committed:
findRecentlyAddedTasks(), findById(), deleteTasks() 수정

findRecentlyAddedTasks(): 파라미터 추가, 인자 검증 로직 추가, unmodifiable view로 반환하도록 수정
 findById(): 코드 간소화
 deleteTasks(): 불필요한 코드 제거

steve7867 committed:
CORS 허용 origin 명시, findRecentlyAddedTasks(), deleteTasks() 수정

findRecentlyAddedTasks(): 쓰레드 안전 로직 추가
deleteTasks(): 반환 타입을 Set<Task>로 변경

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -6,44 +6,46 @@
-    public List<Task> findRecentlyAddedTasks() {
+    public List<Task> findRecentlyAddedTasks(int quantity) {
+        if (quantity <= 0) {
+            throw new IllegalArgumentException("0 또는 음수는 허용되지 않습니다.");
```
문제의 수가 어떤 것인지를 메시지에 포함시키는 것도 괜찮은 방법일 수 있습니다.

steve7867 replied:
추가하겠습니다 ^^

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -6,44 +6,46 @@
+            throw new IllegalArgumentException("0 또는 음수는 허용되지 않습니다.");
+        }
+
         final List<Task> recentlyAddedTaskList = new ArrayList<>();
```
초기화할 때 `new ArrayList` 생성자에 사이즈를 제공해 줄 수도 있겠습니다.

steve7867 replied:
맞습니다.
그럼 내부의 동적 배열을 재할당하고 기존 배열의 요소들을 새로운 배열로 copy하는 일이 없으니 성능상 유리할 것입니다.

steve7867 committed:
findRecentlyAddedTasks() 수정

예외 메시지 수정, list 초기화 size 지정

steve7867 committed:
RecentlyAddedTasksFinder 추가 및 관련 코드 수정

최근에 추가된 Task를 캐시하는 책임을 가진 RecentlyAddedTasksFinder 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -28,10 +28,10 @@ public Collection<Task> findAllTasks() {
     public List<Task> findRecentlyAddedTasks(int quantity) {
         if (quantity <= 0) {
-            throw new IllegalArgumentException("0 또는 음수는 허용되지 않습니다.");
+            throw new IllegalArgumentException(String.format("quantity: %d\n0 또는 음수는 허용되지 않습니다.", quantity));
```
예외 메시지나 로그에 개행 문자(`\n`)를 고의로 넣으면 나중에 로그를 읽을 때 꽤 귀찮아질 수 있습니다.

steve7867 replied:
그렇군요.
제거하겠습니다

johngrib reviewed (approved):
이번주 기본 과제 목표 이상으로 잘 작업하신 것 같습니다.

steve7867 commented:
> 이번주 기본 과제 목표 이상으로 잘 작업하신 것 같습니다.

감사합니다.
개인적으로는 추가 과제를 제대로 구현하지 못해서 많이 아쉽습니다.
내일까지 계속 고민해보겠습니다.

steve7867 committed:
CoreTaskRepository 추가, NullTaskStoredException 추가, RecentlyAddedTasksFinder 삭제

CoreTaskRepository 추가: TaskRepository <- CoreTaskRepository <- TaskRepositoryImpl 순의 상속하는 관계로 구조 변경
RecentlyAddedTasksFinder 삭제: 기존의 RecentlyAddedTasksFinder의 책임을 TaskRepository 로 위임
NullTaskStoredException 추가: 참조값이 null인 Task 객체가 저장되어 있는 경우 사용

steve7867 committed:
TaskController: CORS 허용 Origin 변경, TaskRepositoryImpl 테스트 코드 추가

steve7867 committed:
TaskRepository 메소드 반환 타입 변경 및 관련 코드 리팩토링, findRecentlyAddedTasks() 로직 변경

TaskRepository 메소드 반환 타입 변경: findAllTasks(), findRecentlyAddedTasks()의 반환 타입을 List<Task>로 변경
findRecentlyAddedTasks() 로직 변경: 항상 300개의 task를 캐시하도록 변경

steve7867 committed:
NullTaskStoredException 제거

steve7867 committed:
JavaDoc 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/CoreTaskRepository.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    public Optional<Task> findById(Long id) {
+        if (id <= 0) {
```
null 체크도 함께 할 수 있겠어요.
`if (id == null || id <= 0)`

steve7867 replied:
네 null 체크 로직 추가하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/CoreTaskRepository.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public Set<Task> deleteTasks(Set<Long> idSet) {
```
null 방어 코드도 추가하는 것이 좋겠습니다.

steve7867 replied:
null 체크 로직이 여기저기 다 빠져있네요.
추가하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repositories/TaskRepositoryImpl.java:
```diff
@@ -0,0 +1,128 @@
+        recentlyAddedTasksMap.remove(id);
+        final Optional<Task> deletedTask = super.deleteById(id);
+
+        fillRecentlyAddedTasksMap(cacheSize - getSize());
```
`getSize` 메소드의 이름이 아쉽습니다.

steve7867 replied:
기존의 cacheSize는 cacheCapacity로, 
getSize()는 getCacheSize()로 변경하였습니다.

johngrib reviewed (commented):
2주차라고 믿기지 않을 만큼 작업을 잘 해내신 것 같습니다. 그만큼 열심히 노력하신 것 같아요. 주간 회고를 작성해 공유해 주세요.

steve7867 commented:
> 2주차라고 믿기지 않을 만큼 작업을 잘 해내신 것 같습니다. 그만큼 열심히 노력하신 것 같아요. 주간 회고를 작성해 공유해 주세요.

감사합니다.
하지만 더 좋은 코드를 작성하지 못한 아쉬움은 남네요 ㅎㅎ
다음 주에는 더 열심히 할 수 있도록 하겠습니다.
